### DISCUSSION TITLE
Offline first insert mutation create query of inserted item

### DISCUSSION THREAD

**Discussion by dohomi** [urql-graphql/urql#2038] [Q&A] [answered]
Hello

I follow the https://formidable.com/open-source/urql/docs/graphcache/cache-updates/ page of the docs. Currently I am playing around with an offline first solution and most of the examples show how to update a list in the update method.

But in my scenario (Hasura backend) I have 2 queries which I try to make work in offline mode:
item (list of items)
item-by-pk (one item)

In the create mode I found how to update the list (https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#writing-links-individually) but now I tried several methods on how to create a new query item `item-by-pk` which should be resolve after submitting the data. 

In my code After `insert_item` mutation I redirect to the newly created `item-by-pk` but that query is undefined and does not resolve in offline mode.

Thanks for any hints

**Comment by kitten** [comment, marked as answer]
`item-by-pk` sounds a little like "item by product key", is that right? So, I'd assume `pk` is your key field, right? Because, if that's your key field and a mutation happens to _create_ that item in the cache, then you'll only need to write a resolver that "teaches" Graphcache how that resolution works, e.g. https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#resolving-by-keys

In other words, if we have an `Item` with an `id: ID!` field and that's the key field, then we can define a resolver for `Query.item(id: ID)` that says that this field resolves `(_, args) => ({ __typename: 'Item', id: args.id })`. This of course only works for key fields but I'm wondering whether that's already all you were looking for

**Reply by dohomi** [reply]
@kitten thanks for your answer. Yes item-by-pk is the detail query of an item based on its primary key (ID). I basically have a detailed item-by-pk query with a fragment of the whole item. The list of many items just has a smaller fragment of the item. The result of insert_item has basically all infos for the detail query, I don't see a connection between that result data and the local-resolver function?

**Reply by kitten** [reply]
@dohomi Well, what's important to note is that often updaters are used to write _relations_ only. You'll find for instance that if you update a list that all that actually needs updating is the typebame and keyfields. Essentially, it's easy for Graphcache to store a big table of all objects and their fields — however, it can't automatically know how relations work, it has to see them.

This is why you can send an update and that item will immediately be updated in all queries across the app. But you have to update lists and relations that aren't in your mutation manually 

However, in your case, Graphcache has seen the item in the mutation, and it will store that and its fields automatically. The only thing it's missing is the `Query.item_by_pk(pk: "...")` to `Item` relation, since it won't have seen that field yet.

When you create a resolver for it though, you're telling Graphcache what to do when it first sees this field. This means it can find the object (with all its fields from your mutation) without ever having sent the query with this field to the API

**Reply by dohomi** [reply]
@kitten Yes now I finally found the last piece of the puzzle, as you described I had to tell my item_by_pk on every insert mutation how to find that object. This I did with `cache.link()` and now my roundtrips work as expected. Thank you so much for your input
